dev/spark_overview.md

```diff
@@ -0,0 +1,452 @@
+
+Set `spark.executor.cores` and `spark.executor.memory` to numbers that allow creation of your
+desired number of executors. Those are per-executor values.
+
+### Start the Spark processes
+```
+$SPARK_HOME/sbin/start-master.sh
+```
+Start a worker with this command. Give the worker as much memory as you can afford. You can also
+configure this in `spark-env.sh`.
+```
+$SPARK_HOME/sbin/start-worker.sh -m 16g spark://$(hostname):7077
+```
+
+### Spark UI
+Open http://localhost:8080/ in your browser and copy the cluster URL and port. It will be
+something like `spark://$(hostname):7077`.
+
+If you add the `sbin` to your `PATH` environment variable, here is a one-liner:
+```
+$ start-master.sh && start-worker.sh -m 24g spark://$(hostname):7077
+```
+
+Monitor cluster tasks in your browser.
+
+## Stop the Spark processes
+Stop all of the processes when you complete your work.
+```
+$ stop-worker.sh && stop-master.sh
+```
+
+## Installing a Spark Standalone Cluster on an HPC
+This section describes how you can run Spark jobs on any number of HPC compute nodes.
+
+The HPC GitHub repository contains scripts that will create an ephemeral Spark cluster
+on compute nodes that you allocate.
+
+1. The repository as setup instructions that are not currently correct because the branch
+   is still under review. Follow
+   https://github.com/daniel-thom/HPC/blob/apache-spark-clusters/applications/spark/README.md
+   except substitute step 1 with the steps below.
+
+   Secondly, the README has generic instructions to run Spark in a variety of ways. The rest
+   of this section gives specific guidance on how to apply those instructions to use dsgrid with
+   a Spark cluster in an interactive session.
+```
+$ git clone https://github.com/daniel-thom/HPC.git
+$ cd HPC
+$ git fetch origin apache-spark-clusters
+$ git checkout apache-spark-clusters
+```
+
+2. You probably want to choose compute node(s) with fast local storage. This example will work.
+```
+$ salloc -t 01:00:00 -N1 --account=dsgrid --partition=debug --mem=730G
+```
+Notes:
+- Eagle will let you allocate two debug jobs each with two nodes. So, you can use these scripts
+  to create a four-node cluster for one hour.
+- If the debug partition is not too full, you can append `--qos=standby` to the command above
+  and not be charged any AUs.
+
+3. Setup the environment in a directory in `/scratch/$USER`.
+```
+$ export PATH=$PATH:<your-repo-path>/HPC/applications/spark/spark_scripts
+```
+
+4. Select a container valid for use with dsgrid, which currently requires Spark v3.3.x and
+   Python 3.10. The team has validated the container below. It was created with this Dockerfile
+   in dsgrid: `docker/spark/Dockerfile`
+```
+$ create_config.sh -c `/projects/dsgrid/containers/spark_py310.sif`
+```
+
+5. Customize the Spark configuration files in `./conf` as necessary per the HPC instructions.
+
+6. Start the cluster.
+```
+$ start_spark_cluster.sh
+```
+The cluster URL will be `spark://$(hostname):7077`.
```

Comment:
Is this run from a login node or after ssh-ing into a compute node? If the latter, I haven't been told to do that yet.

Reply:
The salloc command above dumps you into the compute node. Is that considered common knowledge? I'll add a note saying that your are there.